### ISSUE TITLE
Support data parallelism with multi-GPUs

### ISSUE THREAD

**Issue by QiJune** [wangkuiyi/gotorch#369]
Data parallelism replicates the model on every device to generates gradients independently and then communicates those gradients at each iteration to keep model replicas consistent.

Following is a survey for support data parallelism in GoTorch.

1. [NCCL](https://developer.nvidia.com/nccl) and [Gloo](https://github.com/facebookincubator/gloo)

NCCL provides Broadcast and AllReduce C APIs, we could wrapper them in Go, and use them directly in GoTorch.

Gloo is another collective communications library, which supports both CPU and GPU.

The GPU performance of NCCL is better than Gloo.

2. [Horovod](https://github.com/horovod/horovod)

Horovod is a distributed deep learning training framework for TensorFlow, Keras, and PyTorch.

Horovod calls NCCL or Gloo underneath.

Horovod does some optimizations on communication. It merges small gradients into several buckets, which reduces the AllReduce kernel launching cost.

Horovod also supports [elastic training](https://horovod.readthedocs.io/en/latest/elastic_include.html). The biggest difference when moving from normal distributed training to elastic training is the need to track and synchronize among the workers as workers are added or removed from the job.

The elastic training depends on the Gloo library. So, the GPU performance may suffer a little.

3. [PyTorch Distributed Package](https://pytorch.org/tutorials/intermediate/ddp_tutorial.html)

PyTorch distributed package is fast than Horovod with PyTorch. 

It does more optimizations, including using buckets, overlapping communication and computation.  Please refer to this [paper](https://arxiv.org/abs/2006.15704) for more details.

An interesting observation also verifies it. People who want to run TensorFlow with AllReduce distributed strategy will choose Horovod, whereas people who want to run PyTorch with AllReduce distributed strategy will choose `torch.DistributedDataParallel`.

**Comment by QiJune** [comment]
[PyTorch DistributedDataParallel and Horovod distributed training benchmarks](https://www.reddit.com/r/MachineLearning/comments/hmgr9g/d_pytorch_distributeddataparallel_and_horovod/)

**Comment by QiJune** [comment]
libtorch provides a thin wrapper for NCCL/Gloo, [ProcessGroupNCCL](https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/ProcessGroupNCCL.hpp).

[Here](https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/example/allreduce.cpp) is an example:

```cpp
#include <c10d/FileStore.hpp>
#include <c10d/ProcessGroupGloo.hpp>

using namespace ::c10d;

int main(int argc, char** argv) {
  int rank = atoi(getenv("RANK"));
  int size = atoi(getenv("SIZE"));
  auto store = std::make_shared<FileStore>("/tmp/c10d_example", size);
  ProcessGroupGloo pg(store, rank, size);

  // Create some tensors
  const auto ntensors = 10;
  std::vector<at::Tensor> tensors;
  for (auto i = 0; i < ntensors; i++) {
    auto x =
        at::ones({1000, 16 * (i + 1)}, at::TensorOptions(at::CPU(at::kFloat)));
    tensors.push_back(x);
  }

  // Kick off work
  std::vector<std::shared_ptr<ProcessGroup::Work>> pending;
  for (auto i = 0; i < ntensors; i++) {
    std::vector<at::Tensor> tmp = {tensors[i]};
    pending.push_back(pg.allreduce(tmp));
  }

  // Wait for work to complete
  for (auto& work : pending) {
    work->wait();
  }
}

```

It has two advantages comparing with NCCL directly:

- ProcessGroupNCCL takes torch Tensors as input
- ProcessGroupNCCL launches allreduce kernel with a separate CUDA stream, which achieves potentially concurrency and better performance.

**Comment by QiJune** [comment]
We want to train a distributed MNIST example, the following is a MVP(Minimum Viable Product) for the target:

- A RecordIODataLoader to read data after static partition.
- Expose [ProcessGroupNCCL](https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/ProcessGroupNCCL.hpp) to Go, so we could apply broadcast/allreduce operations on torch tensors in Go.
- A Launcher to create training processes for each GPU.

After we complete the above things, we could do more optimizations, including:

- support dynamic data partition
- use the hook mechanism to overlap communication and computation
- bucket small gradients into a big tensor to reduce CUDA kernel launching overhead

**Comment by Mddct** [comment]
Any progress for distributed training
